### PR TITLE
DSPy Support For Snowflake LLM's

### PR DESCRIPTION
**Pull request by sfc-gh-alherrera**
Adding support for using Snowflake's Cortex API with DSPy. This includes an implementation of a Snowflake LM as well as a Snowflake RM.

### CODE REVIEW

File: dspy/retrieve/snowflake_rm.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class SnowflakeRM(dspy.Retrieve):
+    """A retrieval module that uses Weaviate to return the top passages for a given query.
+
+    Assumes that a Snowflake table has been created and populated with the following payload:
+        - content: The text of the passage
+
+    Args:
+        snowflake_credentials: connection parameters for initializing Snowlfake client.
```
Comment (arnavsinghvi11) [inline dspy/retrieve/snowflake_rm.py:26]: couple of typos L26, L28, L29

File: pyproject.toml [hunk 2]
```diff
@@ -128,6 +129,7 @@ pytest-mock = "^3.12.0"
 ruff = "^0.3.0"
 black = "^24.2.0"
 pre-commit = "^3.7.0"
+ipykernel = "^6.29.4"
```
Comment (arnavsinghvi11) [inline pyproject.toml:132]: is this addition needed? don't see it used in the changes

File: setup.py [hunk 3]
```diff
@@ -1,44 +1,45 @@
...
-        "Development Status :: 3 - Alpha",	
-        "Intended Audience :: Science/Research",	
-        "License :: OSI Approved :: MIT License",	
-        "Operating System :: POSIX :: Linux",	
-        "Programming Language :: Python :: 3",	
-        "Programming Language :: Python :: 3.8",	
-        "Programming Language :: Python :: 3.9",	
-    ],	
-)	
+        "snowflake": ["snowflake-snowpark-python"],
```
Comment (arnavsinghvi11) [inline setup.py:34]: since this added as an extras (aka dspy-ai[snowflake]), can you add similar logic to how this handled in other RM imports (example with [Milvus](https://github.com/stanfordnlp/dspy/blob/d7dace6b51e1c0899fd798c575262310939f6deb/dspy/retrieve/milvus_rm.py#L14))  in [SnowflakeRM](https://github.com/stanfordnlp/dspy/pull/859/files#diff-d982e89c27806bdc8dca2b257dab30c39f85db9a9dbbbabdb28ae15ddd550cd8R15)

File: dspy/retrieve/snowflake_rm.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+try:
+    from snowflake.snowpark import Session
+    from snowflake.snowpark import functions as snow_fn
+    from snowflake.snowpark.functions import lit, vector_cosine_distance
+    from snowflake.snowpark.types import VectorType
+
+except ImportError:
+    raise ImportError(
+        "The 'snowflake' extra is required to use SnowflakeCortexRM.:"
+        "Install it with `pip install snowflake-snowpark-python`",
```
Comment (arnavsinghvi11) [inline dspy/retrieve/snowflake_rm.py:15]: check comment below

File: setup.py (hunk 3 above)
Comment (arnavsinghvi11) [reply setup.py:34]: and then also update the README to add this as a supported extras dependency - line starting with "For the optional ..."

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Thanks @sfc-gh-alherrera for the PR! left a few comments and seems like there are some merge conflicts to resolve.

File: dsp/modules/snowflake.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        self.client = self._init_cortex(credentials=credentials)
+        self.provider = "Snowflake"
+        self.kwargs = {
+            **self.kwargs,
+            "temperature": 0.7,
+            "max_output_tokens": 1024,
+            "top_p": 1.0,
+            "top_k": 1,
+            **kwargs,
+        }
```
Comment (arnavsinghvi11) [inline dsp/modules/snowflake.py:78]: need to add `self.history: list[dict[str, Any]] = []`

Comment (sfc-gh-alherrera) [reply dsp/modules/snowflake.py:79]: Done

File: dspy/retrieve/snowflake_rm.py (hunk 4 above)
Comment (sfc-gh-alherrera) [reply dspy/retrieve/snowflake_rm.py]: Done

File: setup.py (hunk 3 above)
Comment (sfc-gh-alherrera) [reply setup.py:35]: Done

File: pyproject.toml (hunk 2 above)
Comment (sfc-gh-alherrera) [reply pyproject.toml]: Note needed, removed

File: dspy/retrieve/snowflake_rm.py (hunk 1 above)
Comment (sfc-gh-alherrera) [reply dspy/retrieve/snowflake_rm.py]: addressed

**Comment by sfc-gh-alherrera** [comment]
@arnavsinghvi11 thanks for taking a look. Just took a whole pass on your comments and addressed conflicts

File: docs/api/language_model_clients/Snowflake.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
+---
+sidebar_position: 
+---
+
+# dspy.Snowflake
+
+### Usage
+
+```python
+lm = dspy.Snowflake(model="mixtral-8x7b",credentials=connection_parameters)
```
Comment (arnavsinghvi11) [inline docs/api/language_model_clients/Snowflake.md:10]: i think it would be useful to define some required `connection_parameters` from Snowflake based on the [documentation](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.Session) so users know what to expect when configuring `dspy.Snowflake`.

File: docs/api/retrieval_model_clients/SnowflakeRM.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+SnowflakeRM uses `e5-base-v2` embeddings model by default or any Snowflake Cortex supported embeddings model.
+
+#### Default OpenAI Embeddings
+
+```python
+from dspy.retrieve.snowflake_rm import SnowflakeRM
+import os
+
+connection_parameters = {
+
```
Comment (arnavsinghvi11) [inline docs/api/retrieval_model_clients/SnowflakeRM.md:59]: ah you can add this to the LM documentation to address that comment!

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
left a few comments @sfc-gh-alherrera. 

Can you also resolve the [failing test](https://github.com/stanfordnlp/dspy/actions/runs/9034447491/job/24855395088?pr=859) from the actions on fixing the poetry.lock file? Additionally, can you run `ruff check . --fix-only` and push again to pass the failing test?

File: docs/api/language_model_clients/Snowflake.md (hunk 6 above)
Comment (sfc-gh-alherrera) [reply docs/api/language_model_clients/Snowflake.md:23]: Done. added the same as in the RM docs

File: docs/api/retrieval_model_clients/SnowflakeRM.md (hunk 7 above)
Comment (sfc-gh-alherrera) [reply docs/api/retrieval_model_clients/SnowflakeRM.md:59]: Done!

**Comment by sfc-gh-alherrera** [comment]
thanks for the review @arnavsinghvi11.  Just went through and address your latest round of comments.

**Comment by arnavsinghvi11** [comment]
Thanks @sfc-gh-alherrera !
